toc/rfc/rfc-draft-cnb-lifecycle.md

```diff
@@ -0,0 +1,135 @@
+
+## Summary
+
+[Cloud Native Buildpacks (CNBs)](https://buildpacks.io/), also known as v3 buildpacks, are the current generation of buildpacks and offer some improvements over the v2 buildpacks that CF Deployment currently uses. The Cloud Foundry Foundation already has an implementation of Cloud Native Buildpacks via the [Paketo](https://paketo.io/) project, however these CNBs can't currently be used in CF.
+
+This RFC introduces a new optional lifecycle to Cloud Foundry which enables users to build their applications using Cloud Native Buildpacks.
+
+## Problem
+
+The v2 buildpacks are effectively in maintenance mode and do not receive substantial new features. By not integrating with v3 buildpacks, Cloud Foundry is missing out on new buildpacks (e.g. Java Native and Web Servers CNBs) as well as any new features that are added to the still-actively-developed v3 buildpacks.
+
+## Proposal
+
+- Introduce a new [lifecycle type](https://v3-apidocs.cloudfoundry.org/index.html#lifecycles) `cnb` and its lifecycle data
+- Introduce a new [app lifecycle](https://github.com/cloudfoundry/diego-design-notes#app-lifecycles) called `cnbapplifecycle` which interacts with the [CNB Lifecycle](https://github.com/buildpacks/lifecycle)
+- Reuse [cflinuxfs4](https://github.com/cloudfoundry/cflinuxfs4) as the base for staging and running apps
+- Introduce a new flag to CLI and the [app manifest](https://docs.cloudfoundry.org/devguide/deploy-apps/manifest-attributes.html) to be able to use Cloud Native Buildpacks instead of v2 buildpacks
+
+### Architecture
+
+This will require changes in the following releases:
+
+- CF CLI
+- Cloud Controller
+- Diego
+
+No changes to how Diego runs workloads are necessary to implement this RFC.
+
+- The cli will forward a new app lifecycle type to the cloud controller.
+- The cloud controller will use a new `cnbapplifecycle` for the application.
+
+Affected cloud controller APIs (all that interact with [lifecycles](https://v3-apidocs.cloudfoundry.org/index.html#lifecycles)):
+
+- [apps](https://v3-apidocs.cloudfoundry.org/index.html#apps)
+- [builds](https://v3-apidocs.cloudfoundry.org/index.html#builds)
+- [droplets](https://v3-apidocs.cloudfoundry.org/index.html#droplets)
+- [manifests](https://v3-apidocs.cloudfoundry.org/index.html#manifests)
+
+### Goals
+
+- Establish the latest generation of buildpacks in CF as first-class citizen
+- Increase cohesion and app portability between CF Deployment and [Korifi](https://www.cloudfoundry.org/technology/korifi/), via mutual Paketo integration
+- Increase adoption of Cloud Native Buildpacks
+- Open the door for eventual deprecation of the v2 buildpacks, reducing maintenance costs (v2 buildpack deprecation is NOT included in this RFC)
+- No fundamental changes to the architecture of CF
+  - Result of the staging process will be a droplet
+  - No OCI registry necessary
+  - Reuse cflinuxfs4 as rootfs during build and run
+
+### High Level Implementation
```

Comment:
I never saw a _great_ solution to the VCAP services => CN bindings problem? Is there a way to make this work that isn't a ton of brittle translation tables?

Reply:
The https://github.com/cloudfoundry/community/pull/804 tries to address this point. I would be interested in your feedback. 

FYI @dmikusa as we discussed this during the TOC meeting and feedback is welcome.

Reply:
@rkoster @dsboulder @beyhan Should we add a little subsection referring to https://github.com/buildpacks/libcnb/pull/228, the possibility to do the same for `packit` based Paketo buildpacks and referring to #804 for solving it once and for all?

In short, I would probably write this as
> Buildpacks that need credentials from services should be adapted to read from `VCAP_SERVICES` until #804 removes this requirement.
> **Note**: Buildpacks based on `libpak` or `libcnb` already understand credentials from `VCAP_SERVICES` since [libcnb#228](https://github.com/buildpacks/libcnb/pull/228) has been merged and released.

Reply:
> Ideally a heading would be added discussion what to do with service bindings.

Should it really? It feels a bit out of place because...

a) This is not necessarily `cf` related, but rather buildpack related
b) This feels like an implementation detail as already established in the predecessor rfc ([here](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0017-add-cnbs.md#service-bindings))

What should we mention here? That `cf` does not have to change?

Reply:
@beyhan like https://github.com/cloudfoundry/community/pull/796/files#r1574769369?

Reply:
yes, thanks!

Reply:
> Should it really? It feels a bit out of place because...
Sure now that we have an RFC addressing the problem of servicebinding spec compatibility there is no need to discuss it in this RFC. A link to the service-binding RFC would be nice so that future readers understand how they relate.